### ISSUE TITLE
className prop not working with custom DOM elements in 0.14.0-rc1

### ISSUE THREAD

**Issue by hellosmithy** [facebook/react#4933]
The `className` prop does not appear to get mapped correctly when applied to custom DOM elements in `0.14.0-rc1`.

JSFiddle: https://jsfiddle.net/hellosmithy/5pdujnfq/1/

**Comment by syranide** [comment]
It works if you set `class` instead.

cc @jimfb perhaps?

**Comment by hellosmithy** [comment]
I think the issue is here: https://github.com/facebook/react/blob/v0.14.0-rc1/src/renderers/dom/shared/ReactDOMComponent.js#L637-L641

Any custom element will match the `isCustomComponent` check and skip all the DOM property operations

**Comment by syranide** [comment]
> I'm struggling to understand the use-case for #3067. Surely attributes should work the same for custom elements and standard DOM elements?

I'm not very familiar with all this, but there's a difference between custom tags `<customtag>` and web components `<web-component>`. Web components are much like React components and aren't themselves visible nodes, so the list of DOM attributes don't inherently apply to them (again, just like it makes little sense passing DOM properties to React components), whereas they do for custom tags which just translate to unstyled DOM nodes.

I assume `<custom-tag>` was previously valid too, but as of the introduction of web components that's no longer true? This is my limited understanding of all this anyway, so don't take my word for it!

**Comment by zpao** [comment]
cc @jimfb

**Comment by jimfb** [comment]
Idk, this seems like expected behavior to me.  For standard DOM nodes, we know a priori that there is no `className` attribute, so we can commandeer it for our own purposes.  For web components, there is no such guarantee, so we can't steal the namespace.  In fact, web component authors may prefer className over class for the same reasons we stole it for standard dom components, and to remain compatible, we need to pass it through unmolested.

Another way of thinking about this: For standard DOM nodes, we conceptually have a custom component defined for each node, that happens to take `className`, just as we happen to take `defaultValue` for uncontrolled form elements and happen to use camel case for attributes.  We happen to follow some consistent naming conventions.  The fact that we do most of our markup transformations at the framework level instead of inside react-component definitions is an implementation detail.  Web component authors are free to follow those same conventions (or not), but we need to pass them through without transformations or we'll confuse the web components.

**Comment by hellosmithy** [comment]
I think there's a distinction between custom elements and web components though. Custom elements are one of the building blocks of web components but they are a spec in their own right. There's a use case for creating semantically meaningful custom elements which don't extend default DOM behaviour.

Wouldn't a more consistent approach be to treat all elements the same within React but allow either users to escape attributes somehow or alternatively they have the option to dangerouslySetInnerHTML if they are indeed using web components that need to operate outside the React scope?

**Comment by jimfb** [comment]
@hellosmithy Using a component with a dash *is* the escape.  As per the specification, a node with a dash in it is guaranteed to be a web component and may have arbitrary attributes defined.

**Comment by hellosmithy** [comment]
@jimfb I'm not sure that's a safe assumption to make. My understanding is that a *custom element* must have a dash in it. Custom elements are used in web components. But an element with a dash in it is not necessarily a web component.

https://developer.mozilla.org/en-US/docs/Web/Web_Components/Custom_Elements
"They are part of Web Components but they can also be used by themselves."

**Comment by jimfb** [comment]
Semantics.  Custom element is a subset of the web component specification.  We use the term "web component" loosely, but I believe the statement holds if you replace "web component" with "custom element"; namely that the custom element may define arbitrary attributes and process those attributes using arbitrary rules.

**Comment by hellosmithy** [comment]
@jimfb I think this is a little more than semantics. Either way this change prevents custom elements from being interoperable with core dom elements, when surely there are other options such as escaping special case attributes (`className` etc) for cases where people need to and thus not create a breaking change. Unless there is another way to use custom elements as purely custom elements and still use React props?

**Comment by jimfb** [comment]
You'd have to special case a whole ton of attributes including every possible camelcase attribute that could ever (past, present, or future) occur on a native DOM component, and some of the rules would be very complex/error-prone.  It would not be as simple as escaping a couple special-cased attributes; the rules would be complex and confusing (perhaps even more so than the current attribute whitelisting).

**Comment by hellosmithy** [comment]
I'm suggesting there should be a discussion on the best way to do it, escaping may not be it. One way would be an escape pattern convention e.g. `myPropEscaped`. But I'm also wondering, if you want to create web components that are outside of React's scope surely that is a case for using `dangerouslySetInnerHTML`, I would have thought that the default behaviour should be interoperability otherwise there's no way to apply React props to custom elements, which seems an odd and inconsistent restriction. On the other hand there *are* other ways to pass through web component attributes in those edge cases where you might want to, and in those cases surely you should be explicitly saying "I don't want React to do the default behaviour in this instance".

**Comment by hellosmithy** [comment]
@jimfb @zpao has there been any more thought on this?

Having read the thread at https://github.com/facebook/react/issues/140 I can see why this direction was taken, but I think it raises inconsistencies and will make code harder to reason about. I really do think escaping is the more consistent and flexible option, and forces code to be explicit if it's breaking React convention. Would be interested to hear what your thoughts are.

**Comment by jimfb** [comment]
@hellosmithy I can't see how to make it work in a way that's better than what we currently have.  If you have an escaping solution that you'd like to propose, we'd be happy to take a look.

The constraints are:
 * Must allow people to specify all possible attributes in a case sensitive way.  Users must be able to specify any combination of `class` and `className` (zero, one, or both).  Users must also be able to specify any combination of `onclick` and `onClick` (zero, one, or both).  This is because custom elements can impose arbitrary processing rules on their attributes, and case sensitivity matters.
 * Must be intuitive for users.  Ideally, the first thing people try should work.  If it doesn't work, users must have an obvious "next step" to try so they can fix their code.
 * Must allow us to do basic sanity checks (like preventing/mitigating XSS attacks) in a way that minimizes the impact on performance.
 * Should not require modifications to the JSX syntax, which is already becoming a defacto standard as other languages/editors/IDEs/parsers understand and implement the syntax, so any changes would break all of them.  You can read the specification here: https://facebook.github.io/jsx/
 * Should play nice with #140 and futureproof us from maintaining an ever-growing whitelist as browsers add new elements/attributes.

**Comment by hellosmithy** [comment]
@jimfb @sebmarkbage I'm happy to submit a PR for escaping attrs for web components. However this seems like a separate issue. The changes that have gone through in this release candidate make it difficult to reason about how React will treat different DOM elements in JSX. The onus should be on those wanting to use web components with their own attribute API's to do so in a way that operates safely with existing React behaviour including the way it treats custom elements.

The changes also seem to contradict what is said here: https://github.com/facebook/react/issues/5052#issuecomment-145594782

> "The primary component strategy for React will never be Web Components."

So it seems odd to me that a breaking change is being introduced here to appease those that want to use web components inside their React components.

**Comment by gargantuan** [comment]
This has to get a big thumbs down form me. Custom elements are useful today. Web Components are not and probably never will be. It's frustrating to lose the semantic goodness of custom elements for no practical reason that helps developers today.

**Comment by cosminnicula** [comment]
+1 for Custom Elements support

**Comment by mikeybox** [comment]
The problem here is actually the determination of what a developer intends by using a custom component. Here you're assuming that a custom html tag will become a web-component and you decide non intuitive attribute behaviour. In a lot of cases custom html elements are used to provide meaning to the HTML based components. The fix here would be to have the option of setting the the behaviour. [Here - ReactDOMComponent.js#L637-L641 -](https://github.com/facebook/react/blob/v0.14.0-rc1/src/renderers/dom/shared/ReactDOMComponent.js#L637-L641) You're testing for a custom component and then deciding that it means a need for custom attribute handling. This is wrong and should be in the hands of developers not framework assumptions. it could easily be set in the component before render is called if this component is intended for promoting as a web-component. This way you remove the need for a breaking change and at the same time provide the functionality you want on a developer chosen basis.

**Comment by jimfb** [comment]
**To be clear: Custom elements DO work in React. You must use the `class` attribute instead of `className` because the custom element spec requires that we allow users to specify arbitrary attributes (including `className`) and we need to preserve that functionality for custom elements.**

There is no conflict with Sebastian's comment in https://github.com/facebook/react/issues/5052#issuecomment-145594782.  Web components are NOT our primary use case, but we will still do the best we can to support the use case for anyone who wants to use them.  Same goes for custom elements.  This is why we've added support for them.

If someone would like to propose a solution that is better than the one we've implemented, subject to the [constraints I mentioned above](https://github.com/facebook/react/issues/4933#issuecomment-143836508), we'd be more than happy to take a look.

**Comment by hellosmithy** [comment]
> To be clear: Custom elements DO work in React.

@jimfb they may work in the sense that they can be rendered to the DOM, but they will not work with React attributes. The inconsistency of using `class` sometimes and `className` other times aside, there is a bigger issue here. _None_ of the React events will work such as `onClick`, `onDrag` etc. Effectively you are leaving "React land" by using a custom element.

I think @mikeybox hit the nail on the head when he said this should be in the hands of developers not the framework.

**Comment by gargantuan** [comment]
I'd like to propose a `<AttributeEscape/>` wrapper component. I'd advocate this solution for three reasons.

1. It's declarative
2. No "magic"
3. It's idiomatic React
4. Doesn't break custom-elements
5. Does not need to be part of the react library. 

So how would this work? It would intercept any child elements, create new DOM elements, assign the props unchanged and dangerously set inner HTML. 

You could also create a white/black list feature giving the developer total control over how attributes are handled. 

The beauty of this solution is that is can be a third party solution and does not need to be part of the react library.

**Comment by artisologic** [comment]
I haven't personally used Web Components/Custom Elements alongside React.
However, reading @jimfb:

> Web components are NOT our primary use case

To me, this makes it even more obvious that React should NOT be handling this itself and should instead leave it to developers / third party (something like @gargantuan suggests).

If it is not the primary use case for React, as a developer, if I know that, and I still put myself in a situation where I want to use both at the same time, then I am prepared to go through a few hoops.

I think it would make sense for React to stay as is, and let developers decide what they want to do.

**Comment by mikeybox** [comment]
@jmfb:
> the custom element spec requires that we allow users to specify arbitrary attributes 

I think we all agree that this is correct. A developer should be allowed to pass all parameters and that is fine. The key here is allow and not enforce. In its current state you're not able to choose and that is the problem. Custom HTML elements do not require this behaviour to be used and essentially most developers will not want to lose basic React element behaviour when using a custom element. By it's very nature custom elements can cover a vast range of intended uses. 

The simple conclusion is that yes it is right to enable this *raw attribute mode*. 
But it should most definitely be a developer choice to opt in to the functionality. This can be either a new component type or a Class property setting.

**Comment by jimfb** [comment]
@gargantuan The devil is in the details.  That does not allow custom elements to play as first class citizens.  You've effectively turned React into a templating system at that point.  You are eliminating React's ability to do diffing (because you convert the virtual dom to dangerouslySetInnerHTML), eliminate the ability for the parent components to make decisions based on the children, eliminate the ability to attach refs and event handlers, etc.

**Comment by gargantuan** [comment]
I think I've not clearly explained how this one example proposal would work. I stress that this shouldn't be part of React. You would only put a web-component inside an AttributeEscape component.

That said, I think it's wrong to focus on my half baked proposal and to ignore the wider problem that others have raised. 

1. Is it Reacts responsibility to deal with web components?
2. Should developers have a choice about how custom elements are handled?

In addition, my proposal has two saving graces:

1. It doesn't break anything. 
2. Web Component users are still free to come implement their own solution. 

Finally - I would plead with you to at least deprecate custom elements and issue a warning before releasing this breaking change. I think that would give the wider community the chance to feed back on change that I'm sure will surprise them.

**Comment by jimfb** [comment]
@gargantuan as per my comment above, it means that elements within AttributeEscape will not play by the vdom rules.  This DOES break lots of things.  And it's a huge problem if you have an AttributeEscape with children placed inside.

**Comment by hellosmithy** [comment]
@jimfb the solution suggested by @gargantuan is one option, with it's own merits and considerations.

The point being made here by me and several other commenters is that the RC1 changes related to custom elements also have their pros and cons and should also be justified to the same stringent standards you are requesting for alternative solutions. Several of us have raised concerns related to consistency, backwards compatibility, explicitness, and developer control. Specifically there are concerns about attributes such as `className` and all of the React events such as `onClick`, `onDrag` etc, which will behave in unexpected and non-declarative ways with the proposed RC1 changes.

I think the point being made here is that the proposed changes at https://github.com/jimfb/react/commit/b1db817dc54eb37e05e5dc1b8586efa8e93b0dac require some further consideration.

Given that "the primary component strategy for React will never be Web Components."...

If it were up to me, I would apply the following constraints:

- Must not break existing React behaviour with regards to events and React attributes such as `className` and React events.
- Users must be able to explicitly specify when they want a DOM elements attributes to be controlled outside of React.
- Should be explicit and declarative

Now I appreciate that there has been a vocal request for the ability to support web components, and specifically to allow attributes to be passed through to them. However I would appeal to your better judgement to consider, is this change really the right solution?

Lets not forget, we are not suggesting to add a feature here, but in fact to hold off form hastily adding in a change without due diligence.

**Comment by sebmarkbage** [comment]
I think the primary motivating factor for allowing something is simply to make it easier once you hit an edge case where you need to consume a custom element and wrap it somehow.

We have escape hatches that you're not supposed to need but still might need for practical reasons.

**Comment by hellosmithy** [comment]
> I think the primary motivating factor for allowing something is simply to make it easier once you hit an edge case where you need to consume a custom element and wrap it somehow.

@sebmarkbage I appreciate there are motivating factors, but are the counter points being fully considered here? Consistency, backwards compatibility etc. Declarative code in particular is one of the things that makes React so great. I realise in this instance the issues raised here were not necessarily apparent when the change was made but there are some pretty strong arguments here for seeking a better solution now that they *have* been raised.

**Comment by jimfb** [comment]
@hellosmithy If you have a better solution, please do propose it.  But please consider that we really did spend an enormous amount of time discussing options&alternatives; this was not a spur-of-the-moment change.  We can't hold back all development for an indeterminate amount of time.  At some point, we need to bite the bullet and make changes in order to make React better.  Ultimately it's just code, and there isn't anything that can't be undone if we find something better in the future, but we did do our due diligence.

If you have a **solution** that you think is better, let us know.  We implemented the best thing we've seen to date.

**Comment by hellosmithy** [comment]
@jimfb but the best solution you are referring to breaks React attributes and events on custom elements. Are you saying this was considered and deemed acceptable?

**Comment by jimfb** [comment]
Using native attributes (instead of the react flavors) for custom elements is the best solution we've seen.  Better event support is still planned, but is blocked by a couple of browser bugs that prevent us from supporting them reliably.  Event support is tracked by several other issues, most notably https://github.com/facebook/react/issues/4751.

**Comment by sebmarkbage** [comment]
Yes. We need a way to render the nodes themselves somehow so that nested React children can work correctly.

It is not possible to map props properly. The only thing that we can safely map automatically is attributes. The rest have to be applied imperatively. So we could either allow no props or at least do attributes automatically.

It turns out that a lot of custom elements work with attributes alone. So it seems like a decent trade off.

**Comment by sebmarkbage** [comment]
If you don't use custom elements then you're not affected.

**Comment by gargantuan** [comment]
Why wont you deprecate the current behaviour?

**Comment by Aaronius** [comment]
In our case, we have a React component that renders a custom element that behaves like a checkbox. The custom element comes from a library that is used throughout our company so we're just wrapping it so we can use it more easily in React (mainly exposing certain properties on the React component for setting handlers for custom events that the custom elements trigger). Internally, the custom element uses a native checkbox for various reasons...one being that if it's within a form and the form is submitted that the value gets sent with the form. Primarily, the custom element just applies a nice custom skin. 

I would like all of React's "special properties" (checked, defaultChecked, value, defaultValue, etc.) that React implements for a native checkbox to apply to my React component as well. I've poked around ReactDOMInput to see how I might apply it to my component but it doesn't really look like it's possible...or at least it isn't clear to me how to do so. Anyway, if there were some way to do that it would be tremendously helpful. Otherwise, we end up trying to manually reproduce these same properties and their behaviors on our React components.

Any tips on how to do so would be helpful. Even a "sorry, you just have to manually recreate those properties on your components" would be helpful as well. Thanks!

**Comment by sebmarkbage** [comment]
Yea, you just have to manually recreate those properties on your components. Sorry. There has been some talk about making these special behaviors composites instead of built-ins but that doesn't seem likely short term.

With a custom element you can probably do better and avoid some of the weird special cases that browsers have to deal with.

**Comment by mikeybox** [comment]
The need for providing direct attribute mapping to the element of a custom element has caused issues for a number of React users. I think the main one being simply that custom elements are in themselves not special. This is the view taken in the w3 specification, they're not special and shouldn't be treated as such. The current way of handling custom elements is to change react behaviour and thereby making them a special case which is often counter productive. I realised today while coding that React already has a precedence for dealing with mapping data straight to DOM. This is seen in the style and dangerouslySet attributes. It would then seem to me that the best course would be to again use that way of thinking to provide a dangerouslySetAttributes (it's probably not dangerous its just user managed) attribute that can be provided with an object to be mapped to the element if required. 

for instance 
```js
let customAttrs = {
    hasChildren: true,
    childName: 'Thomas The Tank Engine'
}
<item-preview dangerouslySetAttributes{ customAttrs }/>

Result:

<item-preview data-reactid=".0.0" haschildren="true" childname="Thomas The Tank Engine"></item-preview>

```

I think this would provide the means of passing arbitrary attributes to the element easy but more importantly assumptions wouldn't need to be made as to the purpose of the custom element.

**Comment by gargantuan** [comment]
After working without custom-elements for a few weeks now, I think it's worth pointing out that my initial fears around losing custom-elements have materialised in force. Inspecting the DOM is now a lot more difficult than it was before. It is now riddled with generic elements that make it difficult to scan the hierarchy of a view composed of several components, which themselves contain components. Attempts to mitigate this with custom attributes, e.g. `<div data-role="my-component">...` makes things worse as it's just more noise. Using the className attribute has the same problem. 

Custom elements were a brilliant way to improve the readability of the DOM. Considering React users will spend a large amount of time inspecting the DOM with developer tools, a sane, readable DOM hierarchy has to be considered an important developer friendly feature. That it's been removed for the benefit of the Web Component crowd is incredibly frustrating.

**Comment by spicyj** [comment]
@gargantuan I'm having trouble understanding what you mean, sorry. FWIW, we recommend you use the React dev tools when debugging React apps because they show the React component structure. If they don't fit your workflow for whatever reason, maybe we can fix that instead.

**Comment by gargantuan** [comment]
@spicyj - there are a few issues with the React dev tools

1. Page reloads reset the hierarchy view
2. The select tool switches to the Elements view
4. No CSS inspection
5. Not compatible with Webpack's hot-loading. (Which is an issue with Webpack, but it still undermines the usefulness of the React dev tools)

In practice I'm still doing most of my debugging in the Elements view and the console. The React tool typically means lot of context switching, and now there are no custom-elements, it means the elements views bears no relation to the react view. As a result, I've found that I rarely use the React Dev tools. 

I think that focusing on the React Dev tools to fix the shortcomings introduced by making custom-elements second class citizens would be over engineering.

**Comment by jimfb** [comment]
> I think that focusing on the React Dev tools to fix the shortcomings introduced by making custom-elements second class citizens would be over engineering.

Now I'm really confused as to what you mean.  Second class citizens in React or second class in the browser?  All React does is render the elements, with all the attributes, as they're passed in.  What happens after that point, in the inspector view, is outside React's control.

The issues you listed above sounds like general devtools issues, which should probably be tracked in https://github.com/facebook/react-devtools

**Comment by Aaronius** [comment]
Thanks for the answer to my question, @sebmarkbage.

**Comment by gargantuan** [comment]
@jimfb - Any React attribute (e.g. className) applied to a custom element is now ignored by React. That's what I mean when I say custom-elements are second class citizens. It's no longer possible to write idiomatic JSX with custom elements. 

This was a breaking change that came out of the blue for me and others and did require a sizeable refactor for anyone who had been using custom-elements. Refactors are not a bad thing - but the reasons for using custom-elements remain while their usefulness has been taken away. No one enjoys a refactor that means their codebase get's more cluttered and difficult to read. 

On React Dev tools: They are neither the issue nor the solution to the issue. I have no complaints about React Dev tools because there are ways for me to do my work without them - but they are not the right tool for the way I work. 

This issue is, and always has been - Custom elements were useful for non-Web Component users and now they're not. It's clear that you don't see the merits of custom elements, and It's clear you don't think anything has been lost with this change. But If you're not going to acknowledge the merits other people see, then there is no discussion to be had we will just have to lump it.

**Comment by spicyj** [comment]
@gargantuan I'm still having trouble understanding exactly what your specific complaint is here. I'm not an expert in web components so please forgive me if this should be obvious. Is the issue only that you could previously write

```
<my-x className="foo" />
```

and now must write

```
<my-x class="foo" />
```

instead? Am I misunderstanding completely or just missing something else that has changed?

**Comment by spicyj** [comment]
I can assure you that the changes made to how custom elements work in 0.14 were intended to make them more useful, not less.

**Comment by hellosmithy** [comment]
@spicyj it's not just `className` although that is the original issue that was raised. Any DOM element props such as all event handlers (`onDrag`, `onClick` etc to name a couple) will also no longer work. It seems strange to have these DOM elements treated differently, and especially to do so automatically without the developer explicitly opting out of React handling the props. I actually really like @mikeybox's suggestion of some kind of `setAttributes` or `dangerouslySetAttributes` attr which would be much more explicit and transparent to the developer.

`class` vs `className` is a bit of a hindrance and an inconsistency. But the event handling is particularly bad I feel. I'd much rather a consistent interface for dealing with DOM elements, including custom elements and be able to add events within React e.g adding the same kind of normalized event handling for `onDrag` or `onClick` rather than have React make the decision for me to switch over to completely escaping the props and passing everything through as web component attributes.

If I created some kind of custom image type as a web component for example I might want it's "web component" api to be `src` but I'd still want to control things like drag and click events from within React. In that instance I feel it's much clearer what's going on if it's assumed that attrs are props unless explicitly declared otherwise. The user would explicitly opt out of React controlled props rather than React implicitly making that decision based on something arbitrary such as wether the element has a hyphen in it's name.

E.g.

```
let escapedAttrs = { src: 'http://...' };
return <custom-image onDrag={ this.handleDrag } setAttrs={ escapedAttrs } />;
```

**Comment by jimfb** [comment]
As I mentioned earlier in this thread...
> Better event support is still planned, but was blocked by a couple of browser bugs that prevent us from supporting them reliably. Event support is tracked by several other issues, most notably #4751.

We want to support events on custom elements - that is on our todo list - but there are some nuances regarding custom-element events that need to be considered and handled properly, and some browser bugs/inconsistencies that were making that infeasible.  The current workaround (until events on custom elements can be officially supported by React) is to attach a `ref` and add the event handlers imperatively.

gargantuan, regarding your prior post, you said:
 > @jimfb... It's clear that you don't see the merits of custom elements...

Please keep these comments in check; you weren't in the room when these topics were being discussed, and you couldn't be more wrong on this particular point.  If you were at our developer meetings, you would see that I've been the strongest proponent of custom elements.  You would see that I've been pushing incredibly hard to provide custom element support.  You would see that I'm always pushing to keep React as compliant with the new web standards as is humanly possible.

We make every attempt to listen to all the feedback and continue to improve React to meet everyone's needs, but you need to understand that custom elements are still considered an [experimental technology](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Custom_Elements) and the details are much more nuanced than you might expect.  When browsers start to converge on the semantics, we can re-introduce some of the diffs that fix the rough edges.

**Comment by hellosmithy** [comment]
@jimfb in response to the events support stuff - maybe I'm misunderstanding but that looks like a potential change to the way events work in general. Wether that happens or not I still don't understand why it wouldn't be preferable to have a consistent approach right now for both events and `className` or any other props that may get added in the future for all DOM elements - then allow developers to escape attrs if they want. Has this been considered and decided it wasn't viable?

**Comment by jimfb** [comment]
@hellosmithy Yes, it was considered.  That approach does not meet constraint numbers 2 and 5 from https://github.com/facebook/react/issues/4933#issuecomment-143836508

**Comment by hellosmithy** [comment]
@jimfb hmmm well I guess this comes down to difference of opinion but to me the escaping solution seems both more intuitive and more future-proof.

1. Intuitive in that it is the same behaviour as is used elsewhere with DOM elements. The difference in behaviour based on the hyphenated vs non-hyphenated element names on the other hand seems completely counterintuitive, especially given there was a precedent of it working the former way.

2. More future-proof in that it keeps control with the developer. A single attr would need to be whitelisted eg. `setAttrs` as a straw-man example, but if a user wanted that to use that very same attr for a web component they would still be able to. I can't see how it would cause issues with https://github.com/facebook/react/issues/140.

**Comment by mikeybox** [comment]
@jimfb I'm not sure why having a setAttr prop for attribute mapping wouldn't support all the constraints? It seems like it would be the most likely fit

> Must be intuitive for users. Ideally, the first thing people try should work. If it doesn't work, users must have an obvious "next step" to try so they can fix their code (ie. what happens when an attribute they wanted to use gets rewritten? do they get a useful error message? how do they intuitively know what to try next?).

If you have an attribute not currently supported by the current list of html attributes then the next step would be to find a way to manually set attributes. We already know that we can set HTML it wouldn't be much of a leap to assume theres a function for attributes and hey thats why we have documentation.


> Should play nice with #140 and futureproof us from maintaining an ever-growing whitelist as browsers add new elements/attributes.

theres no need to place an arbitrary behaviour on dashes in tag names. The w3 standard already declares that custom elements are not special in themselves and should be treated as such. In their basic form they're merely naming convention. React already supports the official list of attributes and while it may change and update over time that would be a problem with standard dom behaviour as it is.
The only real solution to that would be to remove the ambiguity between object props and string attributes.

**Comment by jimfb** [comment]
This thread has been going on way too long, and is circling back to points that I've already made previously in the thread.  I don't think it's useful to keep rehashing this issue.

1.  Rendering exactly what the user typed is intuitive because a user can type `className`, see that it literally outputted `className`, and then try `class` (because, well, it appears to be outputting exactly what they typed) and see that it works.  It doesn't take long to realize the pattern that custom elements get outputted directly, but even if you do miss the pattern, you'll quickly arrive at the correct solution through a single trial-and-error attempt.  With an escaping solution, you would type `className`, it would get rewritten to `class`, and you would not have any reasonable way of knowing that the magic escape prop you need to use is called `setAttrs`.
2.  No, the whitelisting would be extensive.  You would need to whitelist `className`, `htmlFor`, `value`, all the other dom attributes that have boolean values and/or otherwise behave in intelligent ways, all the dom properties that would need to be applied as a property instead of an attribute, all the various dom event handlers, etc.

**Comment by mikeybox** [comment]
@jimfb what we're already seeing is that the change between standard DOM elements and custom element behaviour is counter intuitive and thats the point being made. You'd assume currently that using `className` would output `class` because thats what happens everywhere else. Suddenly the behaviour changes because of a dash in the element name which is arbitrary. You're assuming that the use of a dash means I want to create a web component that is going to have special handling of attributes. It's currently counter intuitive which is why it keeps going around in circles. It seems like the extensive whitelisting is a result of the behaviour change added to custom elements where all props are mapped.

It's clear we all care about this stuff because react is probably the coolest framework around and we use it every day. So yeah people are going to get passionate about it because its so important. Whats important especially to us is the use of custom elements as standard html elements using the current 0.13 behaviour that had no prejudice between elements. All were equal. What we're saying is that the non-equality proposed should be done using a setAttr rather than the current behaviour forking which seems wrong.

**Comment by hellosmithy** [comment]
@jimfb I wonder if we're talking at cross-purposes here. If we're circling back I think it's because the points are being disputed which seems reasonable enough. And to echo @mikeybox it's because we care about keeping React the great project that it is.

> Rendering exactly what the user typed is intuitive because a user can type className, see that it literally outputted `className`, and then try `class`

I can't speak for others but I can say that this certainly wasn't intuitive for me as I debugged this after upgrading to React and it seemed like a bug.

> the whitelisting would be extensive. You would need to whitelist `className`, `htmlFor`, `value`, all the other dom attributes that have boolean values and/or otherwise behave in intelligent ways

What we're suggesting is to not make an arbitrary disctinction. So why would there be any extra whitelisting other than the single point of entry for setting escaped attrs? The suggestion is simply that `custom-elem` by default behaves the same as `div` or `span`. In fact the behaviour worked exactly this way before, just without the extra ability to set some kind of escape attr for those that do want to connect to external web-component APIs.

**Comment by jimfb** [comment]
Ok, I think it's time to call it a day.  We've had a pretty good discussion, but this conversation is looping back on its self and is past the point of being productive.  The thread is just rehashing points that have already been made.  I think we fully understand the concerns mentioned here.  We will update the thread when we have new information to provide.

**Comment by braden-qualtrics** [comment]
Any thoughts on whether or not this ([http://stackoverflow.com/questions/37638268/can-i-put-underscores-in-my-html-tagnames](url)) is a reasonable work around for now?

**Comment by jimfb** [comment]
@braden-qualtrics Sounds like a bad idea to me.  You're going to hit all kinds of edge cases that are going to bite you.  Cases that our current design/support of custom elements is specifically designed to help you handle, as described earlier in this thread.

**Comment by braden-qualtrics** [comment]
@jimfb So using underscores in element names might break React/HTML?

**Comment by gaearon** [comment]
I'm closing since this is not a bug, but a different (and now documented) behavior for web components.
Use `class` for them.

**Comment by sebmarkbage** [comment]
This might be related to #7901. If we make a breaking change to switch to properties then this would become className.
